#### Planteamiento del Problema

Se desea predecir la cantidad de puntos que pueda hacer un jugador en un partido, 
a partir de estadísticas personales (del jugador) y colectivas (del equipo en donde juega). 

Imports

In [28]:
#Se importan todas las dependencias necesarias
import pandas as pd
import sqlalchemy
import psycopg2
import sqlite3
import csv
import os
import numpy as np
import scipy.stats
from scipy.stats import skew, kurtosis
from sqlalchemy import create_engine
from dotenv import load_dotenv
from nba_api.stats.endpoints import playergamelogs, teamgamelogs, teamgamelog
from tabulate import tabulate

Obtenemos las estadísticas de los jugadores por partidos en las temporadas 2020-21, 2021-22 y 2022-23

In [2]:
season0_players = playergamelogs.PlayerGameLogs(season_nullable='2020-21') #Consulta para temporada 2020-21
season1_players = playergamelogs.PlayerGameLogs(season_nullable='2021-22') #Consulta para temporada 2021-22
season2_players = playergamelogs.PlayerGameLogs(season_nullable='2022-23') #Consulta para temporada 2022-23
players = pd.concat([season0_players.get_data_frames()[0],season1_players.get_data_frames()[0],season2_players.get_data_frames()[0]], ignore_index=True) #Dataframe de estadísticas de jugadores por partido en las tres temporadas
players.head()

SEASON_YEAR  PLAYER_ID        PLAYER_NAME NICKNAME     TEAM_ID  \
0     2020-21    1626169    Stanley Johnson  Stanley  1610612761   
1     2020-21    1629052     Oshae Brissett    Oshae  1610612754   
2     2020-21     203083     Andre Drummond    Andre  1610612747   
3     2020-21     202685  Jonas Valanciunas    Jonas  1610612763   
4     2020-21    1628981     Bruno Fernando    Bruno  1610612737   

  TEAM_ABBREVIATION           TEAM_NAME     GAME_ID            GAME_DATE  \
0               TOR     Toronto Raptors  0022001079  2021-05-16T00:00:00   
1               IND      Indiana Pacers  0022001079  2021-05-16T00:00:00   
2               LAL  Los Angeles Lakers  0022001072  2021-05-16T00:00:00   
3               MEM   Memphis Grizzlies  0022001070  2021-05-16T00:00:00   
4               ATL       Atlanta Hawks  0022001066  2021-05-16T00:00:00   

       MATCHUP  ... BLKA_RANK  PF_RANK  PFD_RANK  PTS_RANK  PLUS_MINUS_RANK  \
0  TOR vs. IND  ...     15321     4919       795      1773            17520   
1    IND @ TOR  ...         1    16132       795       560             3883   
2    LAL @ NOP  ...         1    21997      2281      7127            11703   
3    MEM @ GSW  ...     15321    10833      2281       787            20143   
4  ATL vs. HOU  ...     20949     4919      6289      9732             3883   

   NBA_FANTASY_PTS_RANK  DD2_RANK  TD3_RANK  WNBA_FANTASY_PTS_RANK  \
0                  2490      1849       143                   1612   
1                  1231         1       143                    808   
2                  2950         1       143                   3790   
3                   774         1       143                    898   
4                  6572      1849       143                   8016   

   AVAILABLE_FLAG  
0               1  
1               1  
2               1  
3               1  
4               1  

[5 rows x 68 columns]

Obtenemos las estadísticas de los equipos por partidos en las temporadas 2020-21, 2021-22 y 2022-23

In [3]:
season0_teams = teamgamelogs.TeamGameLogs(season_nullable='2020-21') #Consulta para temporada 2020-21
season1_teams = teamgamelogs.TeamGameLogs(season_nullable='2021-22') #Consulta para temporada 2021-22
season2_teams = teamgamelogs.TeamGameLogs(season_nullable='2022-23') #Consulta para temporada 2022-23
teams = pd.concat([season0_teams.get_data_frames()[0],season1_teams.get_data_frames()[0],season2_teams.get_data_frames()[0]], ignore_index=True) #Dataframe de estadísticas de equipos por partido en las tres temporadas
teams.head()

SEASON_YEAR     TEAM_ID TEAM_ABBREVIATION          TEAM_NAME     GAME_ID  \
0     2020-21  1610612766               CHA  Charlotte Hornets  0022001080   
1     2020-21  1610612749               MIL    Milwaukee Bucks  0022001068   
2     2020-21  1610612746               LAC        LA Clippers  0022001074   
3     2020-21  1610612743               DEN     Denver Nuggets  0022001076   
4     2020-21  1610612738               BOS     Boston Celtics  0022001073   

             GAME_DATE    MATCHUP WL   MIN  FGM  ...  AST_RANK  TOV_RANK  \
0  2021-05-16T00:00:00  CHA @ WAS  L  48.0   37  ...       938       826   
1  2021-05-16T00:00:00  MIL @ CHI  L  48.0   38  ...      2131       641   
2  2021-05-16T00:00:00  LAC @ OKC  L  48.0   45  ...      2014         2   
3  2021-05-16T00:00:00  DEN @ POR  L  48.0   44  ...      1741        14   
4  2021-05-16T00:00:00  BOS @ NYK  L  48.0   35  ...      1857       641   

   STL_RANK  BLK_RANK  BLKA_RANK  PF_RANK  PFD_RANK  PTS_RANK  \
0       751       801        332      247       786      1180   
1        66       801       1868      376       786      1031   
2       751      1499       2140      154      2113      1031   
3       751      1830       1361     1160      1609       762   
4      1867       490       1361      942      1437      2049   

   PLUS_MINUS_RANK  AVAILABLE_FLAG  
0             1305               1  
1             1373               1  
2             1305               1  
3             1848               1  
4             1238               1  

[5 rows x 57 columns]

Agregamos la columna PTSR (Puntos recibidos) por el equipo en cada partido, temporadas 2020-21, 2021-22 y 2022-23. 

Sería restar los puntos realizados menos la diferencia final de puntaje respecto al otro equipo una vez finalizado el partido.

In [4]:
teams['PTSR'] = (teams['PTS'] - teams['PLUS_MINUS']).astype(int)
teams.head()

SEASON_YEAR     TEAM_ID TEAM_ABBREVIATION          TEAM_NAME     GAME_ID  \
0     2020-21  1610612766               CHA  Charlotte Hornets  0022001080   
1     2020-21  1610612749               MIL    Milwaukee Bucks  0022001068   
2     2020-21  1610612746               LAC        LA Clippers  0022001074   
3     2020-21  1610612743               DEN     Denver Nuggets  0022001076   
4     2020-21  1610612738               BOS     Boston Celtics  0022001073   

             GAME_DATE    MATCHUP WL   MIN  FGM  ...  TOV_RANK  STL_RANK  \
0  2021-05-16T00:00:00  CHA @ WAS  L  48.0   37  ...       826       751   
1  2021-05-16T00:00:00  MIL @ CHI  L  48.0   38  ...       641        66   
2  2021-05-16T00:00:00  LAC @ OKC  L  48.0   45  ...         2       751   
3  2021-05-16T00:00:00  DEN @ POR  L  48.0   44  ...        14       751   
4  2021-05-16T00:00:00  BOS @ NYK  L  48.0   35  ...       641      1867   

   BLK_RANK  BLKA_RANK  PF_RANK  PFD_RANK  PTS_RANK  PLUS_MINUS_RANK  \
0       801        332      247       786      1180             1305   
1       801       1868      376       786      1031             1373   
2      1499       2140      154      2113      1031             1305   
3      1830       1361     1160      1609       762             1848   
4       490       1361      942      1437      2049             1238   

   AVAILABLE_FLAG  PTSR  
0               1   115  
1               1   118  
2               1   117  
3               1   132  
4               1    96  

[5 rows x 58 columns]

Hacemos una nueva consulta a la API con otro endpoint para poder obtener el W_PCT o % de Victorias del equipo

In [5]:
teams_list = teams.TEAM_ID.unique().tolist() #El for será para encontrar el W_PCT de todos los equipos para cada partido de la temporada
teamLog = pd.DataFrame() #inicializamos para que no quede basura

for team in teams_list:
    teamLog0 = teamgamelog.TeamGameLog(season='2020-21',team_id=team) #Consulta para temporada 2020-21
    teamLog1 = teamgamelog.TeamGameLog(season='2021-22',team_id=team) #Consulta para temporada 2021-22
    teamLog2 = teamgamelog.TeamGameLog(season='2022-23',team_id=team) #Consulta para temporada 2022-23
    teamLog3 = pd.concat([teamLog0.get_data_frames()[0], teamLog1.get_data_frames()[0], teamLog2.get_data_frames()[0]],ignore_index=True) #En cada ciclo se forma el dataframe por equipo por partido en las tres temporadas 
    teamLog  = pd.concat([teamLog, teamLog3]) #Se van consolidando en un dataframe las consultas para todos los equipos

In [6]:
teamLog.head()

Team_ID     Game_ID     GAME_DATE      MATCHUP WL   W   L  W_PCT  MIN  \
0  1610612766  0022001080  MAY 16, 2021    CHA @ WAS  L  33  39  0.458  240   
1  1610612766  0022001064  MAY 15, 2021    CHA @ NYK  L  33  38  0.465  265   
2  1610612766  0022001047  MAY 13, 2021  CHA vs. LAC  L  33  37  0.471  240   
3  1610612766  0022000448  MAY 11, 2021  CHA vs. DEN  L  33  36  0.478  240   
4  1610612766  0022001020  MAY 09, 2021  CHA vs. NOP  L  33  35  0.485  240   

   FGM  ...  FT_PCT  OREB  DREB  REB  AST  STL  BLK  TOV  PF  PTS  
0   37  ...   0.917    10    32   42   25    8    5   11  15  110  
1   39  ...   0.750    15    36   51   28    5    2    8  18  109  
2   33  ...   0.789     8    27   35   24    9    1   10  15   90  
3   42  ...   0.850    15    34   49   30    9    5   17  17  112  
4   42  ...   0.722    11    35   46   22    6    8   17  19  110  

[5 rows x 27 columns]

In [7]:
#Se renombran las columnas de con el mismo titulo que en el dataframe de teams, para poder hacer el merge deseado (filtrando por las columnas GAME_ID y TEAM_ID)
teamLog = teamLog.rename(columns={'Game_ID': 'GAME_ID'})
teamLog = teamLog.rename(columns={'Team_ID': 'TEAM_ID'})

In [8]:
#Se hace el merge deseado de W_PCT (% Victorias) filtrando por las columnas GAME_ID y TEAM_ID
merged_df = pd.merge(teams,teamLog, on=['GAME_ID','TEAM_ID'], how='left')
teams['W_PCT'] = merged_df['W_PCT']
teams.head()

SEASON_YEAR     TEAM_ID TEAM_ABBREVIATION          TEAM_NAME     GAME_ID  \
0     2020-21  1610612766               CHA  Charlotte Hornets  0022001080   
1     2020-21  1610612749               MIL    Milwaukee Bucks  0022001068   
2     2020-21  1610612746               LAC        LA Clippers  0022001074   
3     2020-21  1610612743               DEN     Denver Nuggets  0022001076   
4     2020-21  1610612738               BOS     Boston Celtics  0022001073   

             GAME_DATE    MATCHUP WL   MIN  FGM  ...  STL_RANK  BLK_RANK  \
0  2021-05-16T00:00:00  CHA @ WAS  L  48.0   37  ...       751       801   
1  2021-05-16T00:00:00  MIL @ CHI  L  48.0   38  ...        66       801   
2  2021-05-16T00:00:00  LAC @ OKC  L  48.0   45  ...       751      1499   
3  2021-05-16T00:00:00  DEN @ POR  L  48.0   44  ...       751      1830   
4  2021-05-16T00:00:00  BOS @ NYK  L  48.0   35  ...      1867       490   

   BLKA_RANK  PF_RANK  PFD_RANK  PTS_RANK  PLUS_MINUS_RANK  AVAILABLE_FLAG  \
0        332      247       786      1180             1305               1   
1       1868      376       786      1031             1373               1   
2       2140      154      2113      1031             1305               1   
3       1361     1160      1609       762             1848               1   
4       1361      942      1437      2049             1238               1   

   PTSR  W_PCT  
0   115  0.458  
1   118  0.639  
2   117  0.653  
3   132  0.653  
4    96  0.500  

[5 rows x 59 columns]

In [9]:
#Para comprobar que no hayan quedado valores que no hayan hecho math entre columna GAME_ID y TEAM_ID
#Si fuese el caso, se hubiese rellenado con NaN
teams.W_PCT.isna().any()

False

Obtenemos un solo DataFrame con las columnas que necesitamos, se va a consolidar todo en el Dataframe de estadísticas de los jugadores "players"

In [10]:
#Del Dataframe "teams" se obtienen las columnas de Puntos recibidos por el equipo 'PTSR', 'PTS' Puntos realizados por el equipo y W_PCT (%Victorias)
#Se hace el merge deseado de Puntos recibidos por el equipo 'PTSR', 'PTS' Puntos realizados por el equipo y W_PCT (%Victorias); filtrando por las columnas GAME_ID y TEAM_ID
teams = teams.rename(columns={'PTS': 'PTSTeam'}) #Porque en el Dataframe players tambien hay una columna PTS

merged_df2 = pd.merge(players,teams, on=['GAME_ID','TEAM_ID'], how='left')

players[['PTSRTeam','PTSTeam','W_PCTTeam']] = merged_df2[['PTSR','PTSTeam','W_PCT']] #Se crean las nuevas columnas en el Dataframe players de acuerdo al merge realizado.

players.head()

SEASON_YEAR  PLAYER_ID        PLAYER_NAME NICKNAME     TEAM_ID  \
0     2020-21    1626169    Stanley Johnson  Stanley  1610612761   
1     2020-21    1629052     Oshae Brissett    Oshae  1610612754   
2     2020-21     203083     Andre Drummond    Andre  1610612747   
3     2020-21     202685  Jonas Valanciunas    Jonas  1610612763   
4     2020-21    1628981     Bruno Fernando    Bruno  1610612737   

  TEAM_ABBREVIATION           TEAM_NAME     GAME_ID            GAME_DATE  \
0               TOR     Toronto Raptors  0022001079  2021-05-16T00:00:00   
1               IND      Indiana Pacers  0022001079  2021-05-16T00:00:00   
2               LAL  Los Angeles Lakers  0022001072  2021-05-16T00:00:00   
3               MEM   Memphis Grizzlies  0022001070  2021-05-16T00:00:00   
4               ATL       Atlanta Hawks  0022001066  2021-05-16T00:00:00   

       MATCHUP  ... PTS_RANK  PLUS_MINUS_RANK  NBA_FANTASY_PTS_RANK  DD2_RANK  \
0  TOR vs. IND  ...     1773            17520                  2490      1849   
1    IND @ TOR  ...      560             3883                  1231         1   
2    LAL @ NOP  ...     7127            11703                  2950         1   
3    MEM @ GSW  ...      787            20143                   774         1   
4  ATL vs. HOU  ...     9732             3883                  6572      1849   

   TD3_RANK  WNBA_FANTASY_PTS_RANK  AVAILABLE_FLAG  PTSRTeam  PTSTeam  \
0       143                   1612               1       125      113   
1       143                    808               1       113      125   
2       143                   3790               1        98      110   
3       143                    898               1       113      101   
4       143                   8016               1        95      124   

   W_PCTTeam  
0      0.375  
1      0.472  
2      0.583  
3      0.528  
4      0.569  

[5 rows x 71 columns]

In [11]:
#Para comprobar que no hayan quedado valores que no hayan hecho math entre columna GAME_ID y TEAM_ID
#Si fuese el caso, se hubiese rellenado con NaN
print(players.PTSRTeam.isna().any())
print(players.PTSTeam.isna().any())
print(players.W_PCTTeam.isna().any())

False
False
False


Se agrega una nueva columna para determinar si un equipo es local o visitante

In [12]:
def determinar_tipo(data):
    if '@' in data:
        return 'A'  # Si hay '@', entonces es visitante
    elif 'vs' in data:
        return 'H'  # Si hay 'vs', entonces es local
    
players['HOME_AWAY'] = players['MATCHUP'].apply(determinar_tipo) #Nueva columna Home (H), Away (A)

players.head(2)

SEASON_YEAR  PLAYER_ID      PLAYER_NAME NICKNAME     TEAM_ID  \
0     2020-21    1626169  Stanley Johnson  Stanley  1610612761   
1     2020-21    1629052   Oshae Brissett    Oshae  1610612754   

  TEAM_ABBREVIATION        TEAM_NAME     GAME_ID            GAME_DATE  \
0               TOR  Toronto Raptors  0022001079  2021-05-16T00:00:00   
1               IND   Indiana Pacers  0022001079  2021-05-16T00:00:00   

       MATCHUP  ... PLUS_MINUS_RANK  NBA_FANTASY_PTS_RANK  DD2_RANK  TD3_RANK  \
0  TOR vs. IND  ...           17520                  2490      1849       143   
1    IND @ TOR  ...            3883                  1231         1       143   

   WNBA_FANTASY_PTS_RANK  AVAILABLE_FLAG  PTSRTeam  PTSTeam  W_PCTTeam  \
0                   1612               1       125      113      0.375   
1                    808               1       113      125      0.472   

   HOME_AWAY  
0          H  
1          A  

[2 rows x 72 columns]

Removemos las columnas que no se utilizarán en la generación del modelo predictivo.

In [13]:
players.columns

Index(['SEASON_YEAR', 'PLAYER_ID', 'PLAYER_NAME', 'NICKNAME', 'TEAM_ID',
       'TEAM_ABBREVIATION', 'TEAM_NAME', 'GAME_ID', 'GAME_DATE', 'MATCHUP',
       'WL', 'MIN', 'FGM', 'FGA', 'FG_PCT', 'FG3M', 'FG3A', 'FG3_PCT', 'FTM',
       'FTA', 'FT_PCT', 'OREB', 'DREB', 'REB', 'AST', 'TOV', 'STL', 'BLK',
       'BLKA', 'PF', 'PFD', 'PTS', 'PLUS_MINUS', 'NBA_FANTASY_PTS', 'DD2',
       'TD3', 'WNBA_FANTASY_PTS', 'GP_RANK', 'W_RANK', 'L_RANK', 'W_PCT_RANK',
       'MIN_RANK', 'FGM_RANK', 'FGA_RANK', 'FG_PCT_RANK', 'FG3M_RANK',
       'FG3A_RANK', 'FG3_PCT_RANK', 'FTM_RANK', 'FTA_RANK', 'FT_PCT_RANK',
       'OREB_RANK', 'DREB_RANK', 'REB_RANK', 'AST_RANK', 'TOV_RANK',
       'STL_RANK', 'BLK_RANK', 'BLKA_RANK', 'PF_RANK', 'PFD_RANK', 'PTS_RANK',
       'PLUS_MINUS_RANK', 'NBA_FANTASY_PTS_RANK', 'DD2_RANK', 'TD3_RANK',
       'WNBA_FANTASY_PTS_RANK', 'AVAILABLE_FLAG', 'PTSRTeam', 'PTSTeam',
       'W_PCTTeam', 'HOME_AWAY'],
      dtype='object')

In [14]:
#Remocion de columnas
players = players.drop(['WNBA_FANTASY_PTS', 'GP_RANK', 'W_RANK', 'L_RANK', 'W_PCT_RANK',
       'MIN_RANK', 'FGM_RANK', 'FGA_RANK', 'FG_PCT_RANK', 'FG3M_RANK',
       'FG3A_RANK', 'FG3_PCT_RANK', 'FTM_RANK', 'FTA_RANK', 'FT_PCT_RANK',
       'OREB_RANK', 'DREB_RANK', 'REB_RANK', 'AST_RANK', 'TOV_RANK',
       'STL_RANK', 'BLK_RANK', 'BLKA_RANK', 'PF_RANK', 'PFD_RANK', 'PTS_RANK',
       'PLUS_MINUS_RANK', 'NBA_FANTASY_PTS_RANK', 'DD2_RANK', 'TD3_RANK',
       'WNBA_FANTASY_PTS_RANK','NBA_FANTASY_PTS', 'DD2',
       'TD3'], axis = 1)

players.head(2)

SEASON_YEAR  PLAYER_ID      PLAYER_NAME NICKNAME     TEAM_ID  \
0     2020-21    1626169  Stanley Johnson  Stanley  1610612761   
1     2020-21    1629052   Oshae Brissett    Oshae  1610612754   

  TEAM_ABBREVIATION        TEAM_NAME     GAME_ID            GAME_DATE  \
0               TOR  Toronto Raptors  0022001079  2021-05-16T00:00:00   
1               IND   Indiana Pacers  0022001079  2021-05-16T00:00:00   

       MATCHUP  ... BLKA  PF  PFD  PTS  PLUS_MINUS  AVAILABLE_FLAG  PTSRTeam  \
0  TOR vs. IND  ...    1   1    6   24          -8               1       125   
1    IND @ TOR  ...    0   3    6   31          10               1       113   

   PTSTeam  W_PCTTeam  HOME_AWAY  
0      113      0.375          H  
1      125      0.472          A  

[2 rows x 38 columns]

Tomamos el promedio de los últimos 5 partidos para las estadísticas por jugador

In [15]:
#Ordenar el Dataframe de forma ascendente en base a las columnas descritas.
players = players.sort_values(by=['SEASON_YEAR','PLAYER_ID','GAME_DATE'], ascending=[True,True,True])

In [16]:
#Hacemos reset el indice del dataframe en caso de que lo podamos necesitar en orden
players = players.reset_index(drop=True)

In [ ]:
def calcular_promedio_puntos(dataframe, columna, player_id, game_date, n_partidos):
    """
    Calcula el promedio de puntos de un jugador en los N partidos anteriores, excluyendo el partido actual.
    Parámetros:
      dataframe: El dataframe con los datos de los jugadores.
      game_date: El GAME_ID del partido actual.
      player_id: El PLAYER_ID del jugador para el que se quiere calcular el promedio.
      n_partidos: El número de partidos anteriores a considerar.
      columna: La columna que contiene los puntos del jugador.
    Devuelve:
      El promedio de puntos del jugador en los N partidos anteriores.
    """
    # Filtrar el dataframe para los partidos anteriores al actual y del jugador específico
    partidos_anteriores_jugador = dataframe[(dataframe['GAME_DATE'] < game_date) & (dataframe['PLAYER_ID'] == player_id)]
    
    # Si no hay suficientes partidos anteriores para el jugador, tomar los partidos disponibles
    n_partidos_disponibles = min(len(partidos_anteriores_jugador), n_partidos)
    
    # Si hay al menos un partido anterior, calcular el promedio de puntos
    if n_partidos_disponibles > 0:
        promedio_puntos = partidos_anteriores_jugador.tail(n_partidos_disponibles)[columna].mean()
    else:
        # Si no hay partidos anteriores disponibles, tomar el valor del partido actual
        promedio_puntos = dataframe[(dataframe['GAME_DATE'] == game_date) & (dataframe['PLAYER_ID'] == player_id)][columna].iloc[0]
    
    return promedio_puntos

#Lista de columnas para las cuales se obtendrán las estadísticas promediadas
lista_columnas_para_promedio = ['MIN', 'FGM', 'FGA', 'FG_PCT', 'FG3M', 'FG3A', 'FG3_PCT', 'FTM',
       'FTA', 'FT_PCT', 'OREB', 'DREB', 'REB', 'AST', 'TOV', 'STL', 'BLK',
       'BLKA', 'PF', 'PFD', 'PTS', 'PLUS_MINUS', 'PTSRTeam',
       'PTSTeam', 'W_PCTTeam']

#Tomamos una listas de los jugadores y de las fechas de los partidos.
jugadores_ID = players.PLAYER_ID.unique().tolist()
dates = players.GAME_DATE.unique().tolist()


n_partidos = 5 #numero de partidos para sacar el promedio

#Se llama a la funcion para obtener las estadisticas promediadas de todos los jugadores
for columna in lista_columnas_para_promedio:
  for jugador in jugadores_ID:
      for date in dates:
        
        indice = players.index[(players['GAME_DATE']==date) & (players['PLAYER_ID']==jugador)]

        if indice.empty:
                # Si no se encontró un índice válido, pasar a la siguiente fecha
          continue

        promedio = calcular_promedio_puntos(players, columna, jugador, date, n_partidos)

        if promedio is None:
                # Pasar a la siguiente fecha
          continue
              
        players.loc[indice,columna+'_'+'PROM'] = promedio

Guardado de dataframe en un archivo .csv

In [ ]:
#se guarda el dataframe en el archivo .csv llamado "data_complete_raw.csv"
players.to_csv('/workspace/ProyectoDS/data/raw/data_complete_raw.csv', index=False)

Enviamos los registros a una DB.

In [7]:
players = pd.read_csv('/workspaces/ProyectoDS/data/raw/data_complete_raw.csv')

# Se crea una base de datos llamada "NBA.db" que contendrá los datos del archivo .csv guardado previamente
con = sqlite3.connect("NBA.db")
con.execute("DROP TABLE IF EXISTS ESTADISTICAS;")
players.to_sql('ESTADISTICAS',con, index=False)

con.close()

In [8]:
con = sqlite3.connect("NBA.db")
cur = con.cursor()

# -- Obtenemos los 10 máximos anotadores.
cur.execute("""SELECT SEASON_YEAR, PLAYER_NAME, TEAM_NAME,
            PTS, FGM, FGA, FG_PCT,FG3M, FG3A, FG3_PCT
            FROM ESTADISTICAS
            ORDER BY PTS DESC
            LIMIT 10
            ;

            """)

res = cur.fetchall()

for row in res:
    print(f"Temporada: {row[0]}")
    print(f"Nombre: {row[1]}")
    print(f"Equipo: {row[2]}")
    print(f"Puntos: {row[3]}")
    print(f"Tiros convertidos: {row[4]}/{row[5]}")
    print(f"Porcentaje de aciertos: {row[6]}")
    print(f"Tiros de 3: {row[7]}/{row[8]}")
    print(f"Porcentaje tiros de 3: {row[9]}\n\n")

Temporada: 2022-23
Nombre: Damian Lillard
Equipo: Portland Trail Blazers
Puntos: 71
Tiros convertidos: 22/38
Porcentaje de aciertos: 0.579
Tiros de 3: 13/22
Porcentaje tiros de 3: 0.591


Temporada: 2022-23
Nombre: Donovan Mitchell
Equipo: Cleveland Cavaliers
Puntos: 71
Tiros convertidos: 22/34
Porcentaje de aciertos: 0.647
Tiros de 3: 7/15
Porcentaje tiros de 3: 0.467


Temporada: 2020-21
Nombre: Stephen Curry
Equipo: Golden State Warriors
Puntos: 62
Tiros convertidos: 18/31
Porcentaje de aciertos: 0.581
Tiros de 3: 8/16
Porcentaje tiros de 3: 0.5


Temporada: 2020-21
Nombre: Bradley Beal
Equipo: Washington Wizards
Puntos: 60
Tiros convertidos: 20/35
Porcentaje de aciertos: 0.571
Tiros de 3: 7/10
Porcentaje tiros de 3: 0.7


Temporada: 2020-21
Nombre: Jayson Tatum
Equipo: Boston Celtics
Puntos: 60
Tiros convertidos: 20/37
Porcentaje de aciertos: 0.541
Tiros de 3: 5/7
Porcentaje tiros de 3: 0.714


Temporada: 2021-22
Nombre: Kyrie Irving
Equipo: Brooklyn Nets
Puntos: 60
Tiros convertid

In [9]:
# -- Obtenemos los puntos de jugador por temporada (máximos anotadores en una temporada).
cur.execute("""SELECT PLAYER_NAME, TEAM_NAME, SEASON_YEAR,
              SUM(PTS) AS PTS_TOT
              FROM ESTADISTICAS
              GROUP BY PLAYER_NAME, TEAM_NAME, SEASON_YEAR
              ORDER BY PTS_TOT DESC
              LIMIT 10""")

res = cur.fetchall()

for row in res:
        print(f"Nombre: {row[0]}")
        print(f"Equipo: {row[1]}")
        print(f"Temporada: {row[2]}")
        print(f"Puntos totales: {row[3]}\n\n")

con.close()

Nombre: Jayson Tatum
Equipo: Boston Celtics
Temporada: 2022-23
Puntos totales: 2225


Nombre: Joel Embiid
Equipo: Philadelphia 76ers
Temporada: 2022-23
Puntos totales: 2183


Nombre: Trae Young
Equipo: Atlanta Hawks
Temporada: 2021-22
Puntos totales: 2155


Nombre: Luka Doncic
Equipo: Dallas Mavericks
Temporada: 2022-23
Puntos totales: 2138


Nombre: Shai Gilgeous-Alexander
Equipo: Oklahoma City Thunder
Temporada: 2022-23
Puntos totales: 2135


Nombre: DeMar DeRozan
Equipo: Chicago Bulls
Temporada: 2021-22
Puntos totales: 2118


Nombre: Joel Embiid
Equipo: Philadelphia 76ers
Temporada: 2021-22
Puntos totales: 2079


Nombre: Jayson Tatum
Equipo: Boston Celtics
Temporada: 2021-22
Puntos totales: 2046


Nombre: Stephen Curry
Equipo: Golden State Warriors
Temporada: 2020-21
Puntos totales: 2015


Nombre: Nikola Jokic
Equipo: Denver Nuggets
Temporada: 2021-22
Puntos totales: 2004




Análisis Descriptivo Variables Numéricas

In [36]:
columns_exclude = ['PLAYER_ID','TEAM_ID','GAME_ID']
numeric_columns = players.select_dtypes(exclude=['object', 'category']).columns
numeric_columns = [col for col in numeric_columns if col not in columns_exclude]

Se muestra la media, mínimo, máximo, rango, desviación estándar, cuartiles, asimetría, curtosis y moda.
Esto para las variables numéricas.

In [37]:
estadisticas_desc = []
for column_name in numeric_columns:
    mean = players[column_name].mean()
    min = players[column_name].min()
    max = players[column_name].max()
    moda = players[column_name].mode()
    rang = max - min 
    std = players[column_name].std()
    skewness = skew(players[column_name])
    cuartil_25 = np.percentile(players[column_name],25)
    cuartil_50 = np.percentile(players[column_name],50)
    cuartil_75 = np.percentile(players[column_name],75)
    kurtosis = scipy.stats.kurtosis(players[column_name])

    estadisticas_desc.append([column_name, mean, moda, min, max, rang, std, cuartil_25, cuartil_50, cuartil_75, kurtosis, skewness])

headers = ["Columna", "Media", "Moda", "Mínimo", "Máximo", "Rango", "Desviación estándar", "Cuartil 25", "Cuartil 50", "Cuartil 75", "Kurtosis", "Asimetría"]
print(tabulate(estadisticas_desc, headers=headers, tablefmt="grid"))

+-----------------+-------------+-----------+-------------+----------+----------+-----------------------+--------------+--------------+--------------+---------------+-------------+
| Columna         |       Media |      Moda |      Mínimo |   Máximo |    Rango |   Desviación estándar |   Cuartil 25 |   Cuartil 50 |   Cuartil 75 |      Kurtosis |   Asimetría |
+=================+=============+===========+=============+==========+==========+=======================+==============+==============+==============+===============+=============+
| MIN             |  22.8053    |  12       |   0         |  56.5167 |  56.5167 |            10.7069    |    15.1833   |      23.7383 |      31.5133 |   -0.805723   |  -0.278994  |
+-----------------+-------------+-----------+-------------+----------+----------+-----------------------+--------------+--------------+--------------+---------------+-------------+
| FGM             |   3.89681   |   1       |   0         |  22      |  22      |             3

/home/vscode/.local/lib/python3.10/site-packages/tabulate/__init__.py:827: FutureWarning: Calling float on a single element Series is deprecated and will raise a TypeError in the future. Use float(ser.iloc[0]) instead
  conv(string)
/home/vscode/.local/lib/python3.10/site-packages/tabulate/__init__.py:1232: FutureWarning: Calling float on a single element Series is deprecated and will raise a TypeError in the future. Use float(ser.iloc[0]) instead
  return format(float(val), floatfmt)


Dentro de las apreciaciones iniciales se puede observar que:

-   MIN: 
    - Los jugadores juegan una media de 22.8 minutos por partido.
    - La cantidad de minutos jugados por partido que más se repite es de 12 minutos.
    - La mayor cantidad de minutos que jugó un jugador en un partido fue de 56.5 minutos.

-   FGM: 
    - Los jugadores aciertan una media de aprox. 4 tiros de campo por partido. 
    - La mayor cantidad de tiros de campos acertados en un partido fue de 22.
    - La cantidad de tiros de campo acertados por partido que más se repite es de 1.

-   FTM:
    - Los jugadores aciertan una media de aprox. 2 tiros libres por partido. 
    - La mayor cantidad de tiros libres acertados en un partido fue de 23.

-   PF y PFD:
    - Los jugadores cometen y reciben una media de aprox. 2 faltas por partido. 
    - La cantidad de faltas cometidas por partido que más se repite es de 1.
    - La mayor cantidad de fatas recibidas por un jugador en un partido fue de 19.

-   PTS:
    - La mayor cantidad de puntos anotadas por un jugador en un partido fue de 71 puntos.
    - Los jugadores anotan una media de aprox. 11 puntos por partido.
    
-   PTSTeam y PTSRTeam:
    - La mayor cantidad de puntos hechos y recibidos por equipos en un partido fue de 176.
    - La menor cantidad de puntos hechos y recibidos por equipos en un partido fue de 73.
    - Los equipos hacen y reciben una media de 113 puntos por partido.

Analizamos las variables categóricas

In [84]:
categ_columns = players.select_dtypes(include=['object','category']).columns

categ_columns = categ_columns.tolist()
categ_columns.append('AVAILABLE_FLAG')
categ_columns = pd.Index(categ_columns)
categ_columns

Index(['SEASON_YEAR', 'PLAYER_NAME', 'NICKNAME', 'TEAM_ABBREVIATION',
       'TEAM_NAME', 'GAME_DATE', 'MATCHUP', 'WL', 'HOME_AWAY',
       'AVAILABLE_FLAG'],
      dtype='object')

In [63]:
players.AVAILABLE_FLAG.value_counts()

AVAILABLE_FLAG
1    74969
2       18
Name: count, dtype: int64